File: f#/src/SodiumFRP.FSharp/_Root/Transaction.fs
namespace rec SodiumFRP.FSharp

open System
open System.Collections.Generic
open System.Linq
open System.Runtime.ExceptionServices
open System.Threading
open Priority_Queue

module internal TransactionPrivate =
    let localTransaction = new ThreadLocal<Transaction option>(fun () -> None)
    let transactionLock = obj()
    let onStartHooks = List<unit -> unit>()

type internal ApplyResult<'T> = | Result of 'T | Exception of exn
type internal TransactionDelegate = delegate of Transaction -> unit

type Transaction =
    [<DefaultValue>] static val mutable private runningOnStartHooks : bool
    val mutable private isElevated : bool
    val mutable private entries : List<Entry>
    val private sendQueue : List<Transaction -> unit>
    val mutable private sampleQueue : List<unit -> unit>
    val private lastQueue : Queue<unit -> unit>
    val private postQueue : Queue<Transaction -> unit>
    val mutable private splitQueue : Dictionary<int, Transaction -> unit>
    val private hasParentTransaction : bool
    val mutable internal targetsToActivate : List<Target>
    val mutable internal activatedTargets : bool
    val private prioritizedQueue : SimplePriorityQueue<Entry, int64>
    val mutable private toRegen : bool
    
    private new
        (
            postQueue : Queue<Transaction -> unit>,
            splitQueue : Dictionary<int, Transaction -> unit>,
            hasParentTransaction : bool) =
        {
            isElevated = false;
            entries = List<Entry>();
            sendQueue = List<Transaction -> unit>();
            sampleQueue = List<unit -> unit>();
            lastQueue = Queue<unit -> unit>();
            postQueue = postQueue;
            splitQueue = splitQueue;
            hasParentTransaction = hasParentTransaction;
            targetsToActivate = List<Target>();
            activatedTargets = false;
            prioritizedQueue = SimplePriorityQueue<Entry, int64>();
            toRegen = false
        }
    
    internal new() = Transaction(Queue<Transaction -> unit>(), Dictionary<int, Transaction -> unit>(), false)
    
    static member internal HasCurrentTransaction () = TransactionPrivate.localTransaction.Value.IsSome
    static member internal GetCurrentTransaction () = TransactionPrivate.localTransaction.Value
    
    static member private RunStartHooks () =
        if not Transaction.runningOnStartHooks then
            Transaction.runningOnStartHooks <- true
            try for action in TransactionPrivate.onStartHooks do action ()
            finally Transaction.runningOnStartHooks <- false
    
    static member private EnsureElevated (t : Transaction) =
        if not t.isElevated then
            t.isElevated <- true
            if not t.hasParentTransaction then
                Monitor.Enter TransactionPrivate.transactionLock
            Transaction.RunStartHooks ()
    
    static member internal Apply (code : Transaction -> bool -> 'T) (ensureElevated : bool) =
        let transaction = TransactionPrivate.localTransaction.Value
        let createdNewTransaction = transaction.IsNone
        let (result, newTransaction) =
            try
                let newTransaction = transaction |> Option.defaultWith Transaction
                if createdNewTransaction then TransactionPrivate.localTransaction.Value <- Some newTransaction
                if ensureElevated then Transaction.EnsureElevated newTransaction
                (Result (code newTransaction createdNewTransaction), Some newTransaction)
            with | e -> (Exception e, None)
        
        try
            try if createdNewTransaction then newTransaction |> Option.iter (fun t -> t.Close ())
            with
                | e2 ->
                    match result with
                        | Result _ -> reraise ()
                        | Exception e -> AggregateException(e, e2) |> raise
            
            match result with
                | Result r -> r
                | Exception e ->
                    (ExceptionDispatchInfo.Capture e).Throw ()
                    raise e
        finally
            if createdNewTransaction then
                newTransaction |> Option.iter (fun t ->
                    if t.isElevated && not t.hasParentTransaction then
                        Monitor.Exit TransactionPrivate.transactionLock)
                TransactionPrivate.localTransaction.Value <- None
    
    member internal this.Send a = this.sendQueue.Add a
    
    member internal this.Prioritized (node : Node) a =
        let e = Entry(node, a)
        lock Node.NodeRanksLock (fun () -> this.prioritizedQueue.Enqueue(e, node.Rank))
        this.entries.Add e
    
    member internal this.Sample a = this.sampleQueue.Add a
    
    member internal this.Last a = this.lastQueue.Enqueue a
    
    member internal this.Post a = this.postQueue.Enqueue a
    
    member internal this.Split index a =
        let found, existing = this.splitQueue.TryGetValue index
        let n = if found then Delegate.Combine(TransactionDelegate(existing), TransactionDelegate(a)) :?> TransactionDelegate else TransactionDelegate(a)
        this.splitQueue.[index] <- n.Invoke
    
    member internal this.SetNeedsRegenerating () = this.toRegen <- true
    
    member private this.CheckRegen () =
        if this.toRegen then
            this.toRegen <- false
            this.prioritizedQueue.Clear ()
            lock Node.NodeRanksLock (fun () ->
                let newEntries = List<Entry>(this.entries.Count)
                for e in this.entries do
                    if not e.IsRemoved then
                        newEntries.Add e
                        this.prioritizedQueue.Enqueue(e, e.Node.Rank)
                this.entries <- newEntries)
    
    member private this.Close () =
        Transaction.EnsureElevated this
        for target in this.targetsToActivate do target.IsActivated <- true
        this.activatedTargets <- true
        for i = 0 to this.sendQueue.Count - 1 do this.sendQueue.[i] this
        this.sendQueue.Clear ()
        while this.prioritizedQueue.Count > 0 || this.sampleQueue.Count > 0 do
            while this.prioritizedQueue.Count > 0 do
                this.CheckRegen ()
                let e = this.prioritizedQueue.Dequeue ()
                e.IsRemoved <- true
                e.Action this
            let sq = this.sampleQueue
            this.sampleQueue <- List<unit -> unit>()
            for s in sq do s ()
        while this.lastQueue.Count > 0 do this.lastQueue.Dequeue () ()
        if not this.hasParentTransaction then
            let executeInNewTransaction a runStartHooks =
                try
                    let transaction = Transaction(this.postQueue, this.splitQueue, true)
                    if not runStartHooks then transaction.isElevated <- true
                    TransactionPrivate.localTransaction.Value <- Some transaction
                    try a transaction finally transaction.Close ()
                finally TransactionPrivate.localTransaction.Value <- Some this
            
            while this.postQueue.Count > 0 || this.splitQueue.Count > 0 do
                while this.postQueue.Count > 0 do executeInNewTransaction (this.postQueue.Dequeue ()) true
                let sq = this.splitQueue
                this.splitQueue <- Dictionary<int, Transaction -> unit>()
                for n in sq.Keys.OrderBy(fun v -> v) do executeInNewTransaction sq.[n] false

type internal Entry =
    val public Node : Node
    val public Action : Transaction -> unit
    val mutable public IsRemoved : bool
    new(node, a) = { Node = node; Action = a; IsRemoved = false }

type [<AbstractClass>] internal Node(rank : int64) =
    static let listenersLock = obj()
    
    static let rec ensureBiggerThan (node : Node) limit =
        let rec ensureBiggerThan (originalNode : Node) (node : Node) limit =
            if LanguagePrimitives.PhysicalEquality originalNode node then failwith "A dependency cycle was detected."
            if node.Rank <= limit then
                node.Rank <- (limit + 1L)
                for n in node.GetListenerNodesUnsafe () do ensureBiggerThan originalNode n node.Rank
        if node.Rank > limit then false
        else
            node.Rank <- limit + 1L
            lock listenersLock (fun () ->
                for n in node.GetListenerNodesUnsafe () do ensureBiggerThan node n node.Rank)
            true

    static member val Null = Node<_>(Int64.MaxValue)
    
    static member ListenersLock = listenersLock
    static member val NodeRanksLock = obj()
    static member EnsureBiggerThan node limit = ensureBiggerThan node limit

    member val Rank = rank with get,set

    abstract member GetListenerNodesUnsafe : unit -> Node list

type [<AbstractClass>] internal Target =
    val public Node : Node
    val mutable public IsActivated : bool
    new(node, isActivated) = { Node = node; IsActivated = isActivated }

type internal Node<'T> internal (rank : int64) =
    inherit Node(rank)

    let mutable listeners = HashSet<_>()
    let mutable listenersCapacity = 0
    
    internal new() = Node<_>(0L)

    member this.Link (transaction : Transaction) a target =
        let t = Target<_>(a, target, transaction.activatedTargets)
        if not transaction.activatedTargets then transaction.targetsToActivate.Add t
        lock Node.ListenersLock (fun () ->
            listeners.Add t |> ignore
            listenersCapacity <- listenersCapacity + 1)
        let changed = lock Node.NodeRanksLock (fun () -> Node.EnsureBiggerThan target this.Rank)
        (changed, t)

    member this.Unlink target = this.RemoveListener target
    
    member __.GetListenersCopy () = lock Node.ListenersLock (fun () -> List.ofSeq listeners)
    
    member __.RemoveListener l =
        lock Node.ListenersLock (fun () ->
            listeners.Remove l |> ignore
            if listenersCapacity > 100 && listeners.Count < listenersCapacity / 2 then
                listeners <- HashSet<_>(listeners)
                listenersCapacity <- listeners.Count)

    override __.GetListenerNodesUnsafe () = Seq.map (fun (t : Target<'T>) -> t.Node) listeners |> List.ofSeq

type internal Target<'T> =
    inherit Target
    val public Action : WeakReference<Transaction -> 'T -> unit>
    new (a, node, isActivated) =
        { inherit Target(node, isActivated); Action = new WeakReference<Transaction -> 'T -> unit>(a) }

module Transaction =
    let isActive () = Transaction.HasCurrentTransaction()
    
    let run f = Transaction.Apply (fun _ _ -> f ()) false
    
    let onStart a = lock TransactionPrivate.transactionLock (fun () -> TransactionPrivate.onStartHooks.Add a)
    
    let post a =
        let applyPost (t : Transaction) createdNewTransaction =
            if createdNewTransaction then a () else t.Post (fun _ -> a ())
        Transaction.Apply applyPost false
    
    let internal prioritized node a (transaction : Transaction) = transaction.Prioritized node a

File: dotnet/src/Sodium/Sodium.Core.Frp/Transaction.cs
using System;
using System.Collections.Generic;
using System.Runtime.ExceptionServices;
using System.Threading;

namespace Sodium.Frp
{
    /// <summary>
    ///     A class for managing transactions.
    /// </summary>
    internal sealed class TransactionInternal
    {
        // [ThreadStatic] rather than ThreadLocal<T>: these are read on essentially every public
        // entry point, and a thread-static field is a direct TLS access where ThreadLocal<T>.Value
        // goes through a generic slot table. Nothing here uses ThreadLocal's extra surface
        // (Values, IsValueCreated, value factories, disposal), so the two are interchangeable.
        [ThreadStatic]
        private static TransactionInternal localTransaction;

        [ThreadStatic]
        private static bool runningOnStartHooks;

        // Coarse-grained lock that's held during the whole transaction.
        //
        // Serializing every transaction process-wide is a deliberate guarantee of the library, not an
        // implementation artifact: it is what makes a transaction atomic across threads and keeps update
        // ordering deterministic, which is why callers need no synchronization of their own. Narrowing or
        // sharding this lock would change the library's threading semantics, however tempting it looks as a
        // contention fix. The guarantee and its consequences are documented in the remarks on the public
        // Sodium.Frp.Transaction class.
        private static readonly object TransactionLock = new object();

        private bool isElevated;
        private bool obtainedLock;
        internal int InCallback;
        private static readonly List<Action> OnStartHooks = new List<Action>();
        private readonly List<Action<TransactionInternal>> sendQueue = new List<Action<TransactionInternal>>();
        private List<Action> sampleQueue = new List<Action>();
        private readonly Queue<Action> lastQueue = new Queue<Action>();
        private readonly Queue<Action<TransactionInternal>> postQueue;
        private Dictionary<int, Action<TransactionInternal>> splitQueue;
        private readonly bool hasParentTransaction;
        internal readonly List<Node.Target> TargetsToActivate;
        internal bool ActivatedTargets;

        private static readonly EntryPriorityQueue prioritizedQueue = new EntryPriorityQueue();

        public readonly HashSet<Entry> RerankEntriesSet = new HashSet<Entry>();

        internal TransactionInternal()
            : this(new Queue<Action<TransactionInternal>>(), new Dictionary<int, Action<TransactionInternal>>(), false)
        {
        }

        private TransactionInternal(
            Queue<Action<TransactionInternal>> postQueue,
            Dictionary<int, Action<TransactionInternal>> splitQueue,
            bool hasParentTransaction = true)
        {
            this.postQueue = postQueue;
            this.splitQueue = splitQueue;
            this.hasParentTransaction = hasParentTransaction;
            this.TargetsToActivate = new List<Node.Target>();
        }

        internal static bool IsActiveImpl() => HasCurrentTransaction();

        /// <summary>
        ///     Return whether there is a current transaction.
        /// </summary>
        /// <returns><code>true</code> if there is a current transaction, <code>false</code> otherwise.</returns>
        internal static bool HasCurrentTransaction() => localTransaction != null;

        /// <summary>
        ///     Return the current transaction or <code>null</code>.
        /// </summary>
        /// <returns>The current transaction or <code>null</code>.</returns>
        internal static TransactionInternal GetCurrentTransaction() => localTransaction;

        internal static T RunImpl<T>(Func<T> f) => Apply((_, __) => f());

        internal static T Apply<T>(Func<TransactionInternal, bool, T> code)
        {
            TransactionInternal transaction = localTransaction;

            T returnValue = default;
            Exception exception = null;
            TransactionInternal newTransaction = transaction;
            try
            {
                bool createdNewTransaction = newTransaction == null;
                if (newTransaction == null)
                {
                    newTransaction = new TransactionInternal();

                    localTransaction = newTransaction;
                }

                EnsureElevated(newTransaction);

                returnValue = code(newTransaction, createdNewTransaction);
            }
            catch (Exception e)
            {
                exception = e;
            }

            try
            {
                try
                {
                    if (transaction == null)
                    {
                        newTransaction?.Close();
                    }
                }
                catch (Exception e)
                {
                    if (exception == null)
                    {
                        throw;
                    }

                    throw new AggregateException(exception, e);
                }

                if (exception != null)
                {
                    ExceptionDispatchInfo.Capture(exception).Throw();
                }

                return returnValue;
            }
            finally
            {
                if (transaction == null)
                {
                    if (newTransaction != null && newTransaction.obtainedLock)
                    {
                        Monitor.Exit(TransactionLock);
                    }

                    localTransaction = null;
                }
            }
        }

        private static void EnsureElevated(TransactionInternal transaction)
        {
            if (transaction != null && !transaction.isElevated)
            {
                transaction.isElevated = true;

                if (!runningOnStartHooks)
                {
                    if (!transaction.hasParentTransaction)
                    {
                        Monitor.Enter(TransactionLock);
                        transaction.obtainedLock = true;
                    }

                    RunStartHooks(transaction);
                }
            }
        }

        internal static void OnStartImpl(Action action)
        {
            lock (TransactionLock)
            {
                OnStartHooks.Add(action);
            }
        }

        private static void RunStartHooks(TransactionInternal transaction)
        {
            if (OnStartHooks.Count > 0)
            {
                try
                {
                    localTransaction = null;
                    runningOnStartHooks = true;

                    foreach (Action action in OnStartHooks)
                    {
                        action();
                    }
                }
                finally
                {
                    localTransaction = transaction;
                    runningOnStartHooks = false;
                }
            }
        }

        internal void Send(Action<TransactionInternal> action) => this.sendQueue.Add(action);

        internal void Prioritized(Node node, Action<TransactionInternal> action) =>
            this.Prioritized(new ActionEntry(node, action));

        internal void Prioritized(Entry e)
        {
            lock (Node.NodeRanksLock)
            {
                prioritizedQueue.Enqueue(e);
            }
        }

        internal void Sample(Action action) => this.sampleQueue.Add(action);

        /// <summary>
        ///     Add an action to run after all prioritized actions.
        /// </summary>
        /// <param name="action">The action to run after all prioritized actions.</param>
        internal void Last(Action action) => this.lastQueue.Enqueue(action);

        /// <summary>
        ///     Add an action to run after all last actions.
        /// </summary>
        /// <param name="action">The action to run after all last actions.</param>
        internal UnitInternal Post(Action<TransactionInternal> action)
        {
            this.postQueue.Enqueue(action);

            return UnitInternal.Value;
        }

        /// <summary>
        ///     Add an action to run after all last actions.
        /// </summary>
        /// <param name="index">The order index in which to run the action.</param>
        /// <param name="action">The action to run after all last actions.</param>
        internal UnitInternal Split(int index, Action<TransactionInternal> action)
        {
            // If an entry exists already, combine the old one with the new one.
            Action<TransactionInternal> @new;
            if (this.splitQueue.TryGetValue(index, out Action<TransactionInternal> existing))
            {
                @new = existing + action;
            }
            else
            {
                @new = action;
            }

            this.splitQueue[index] = @new;

            return UnitInternal.Value;
        }

        internal static void PostImpl(Action action)
        {
            // -1 will mean it runs before anything split/deferred, and will run
            // outside a transaction context.
            Apply(
                (trans, createdNewTransaction) =>
                {
                    if (createdNewTransaction)
                    {
                        action();
                    }
                    else
                    {
                        trans.Post(_ => action());
                    }

                    return UnitInternal.Value;
                });
        }

        // If the priority queue has entries in it when we modify any of the nodes'
        // ranks, then we need to re-generate it to make sure it's up-to-date.
        private void CheckRegen()
        {
            foreach (Entry entry in this.RerankEntriesSet)
            {
                prioritizedQueue.ChangeRank(entry, entry.Node.Rank);
            }

            this.RerankEntriesSet.Clear();
        }

        internal void Close()
        {
            try
            {
                EnsureElevated(this);

                foreach (Node.Target target in this.TargetsToActivate)
                {
                    target.IsActivated = true;
                }

                this.ActivatedTargets = true;

                // ReSharper disable once ForCanBeConvertedToForeach
                for (int i = 0; i < this.sendQueue.Count; i++)
                {
                    this.sendQueue[i](this);
                }

                this.sendQueue.Clear();

                while (!prioritizedQueue.IsEmpty() || this.sampleQueue.Count > 0)
                {
                    while (!prioritizedQueue.IsEmpty())
                    {
                        this.CheckRegen();

                        Entry e = prioritizedQueue.Dequeue();
                        e.IsRemoved = true;
                        e.Execute(this);
                        e.Dispose();
                    }

                    List<Action> sq = this.sampleQueue;
                    this.sampleQueue = new List<Action>();
                    foreach (Action s in sq)
                    {
                        s();
                    }
                }

                while (this.lastQueue.Count > 0)
                {
                    this.lastQueue.Dequeue()();
                }

                if (!this.hasParentTransaction)
                {
                    void ExecuteInNewTransaction(Action<TransactionInternal> action, bool runStartHooks)
                    {
                        try
                        {
                            TransactionInternal transaction = new TransactionInternal(this.postQueue, this.splitQueue);

                            if (!runStartHooks)
                            {
                                // this will ensure we don't run start hooks
                                transaction.isElevated = true;
                            }

                            localTransaction = transaction;
                            try
                            {
                                action(transaction);
                            }
                            finally
                            {
                                transaction.Close();
                            }
                        }
                        finally
                        {
                            localTransaction = this;
                        }
                    }

                    while (this.postQueue.Count > 0 || this.splitQueue.Count > 0)
                    {
                        while (this.postQueue.Count > 0)
                        {
                            ExecuteInNewTransaction(this.postQueue.Dequeue(), true);
                        }

                        Dictionary<int, Action<TransactionInternal>> sq = this.splitQueue;
                        this.splitQueue = new Dictionary<int, Action<TransactionInternal>>();

                        List<int> splitIndexes = new List<int>(sq.Keys);
                        splitIndexes.Sort();
                        foreach (int n in splitIndexes)
                        {
                            ExecuteInNewTransaction(sq[n], false);
                        }
                    }
                }
            }
            catch
            {
                this.sendQueue.Clear();
                
                while (!prioritizedQueue.IsEmpty())
                {
                    Entry e = prioritizedQueue.Dequeue();
                    e.IsRemoved = true;
                    e.Dispose();
                }
                
                this.sampleQueue = new List<Action>();
                
                this.lastQueue.Clear();
                
                this.postQueue.Clear();
                
                this.splitQueue = new Dictionary<int, Action<TransactionInternal>>();
                
                throw;
            }
        }

        internal abstract class Entry : IDisposable
        {
            public readonly Node Node;
            public bool InPq;
            public int PqRank;
            public Entry PqNext;
            public Entry PqPrev;
            public bool IsRemoved;

            // Where this entry sits in Node.Entries, so removal needs neither a search nor a
            // shift. -1 means "not in the list", which also makes a second Dispose a no-op
            // rather than removing whatever entry now occupies the old position.
            private int nodeEntryIndex;

            protected Entry(Node node)
            {
                this.Node = node;
                this.PqRank = node.Rank;
                this.nodeEntryIndex = node.Entries.Count;
                node.Entries.Add(this);
            }

            // Subclasses carry whatever state the queued work needs as fields, so a caller on a
            // hot path can avoid allocating a closure and a delegate on top of the entry itself.
            public abstract void Execute(TransactionInternal trans);

            public void Dispose()
            {
                int index = this.nodeEntryIndex;
                if (index < 0)
                {
                    return;
                }

                this.nodeEntryIndex = -1;

                // Swap the last entry into this slot rather than shifting everything after it
                // down. Node.Entries is only ever walked to collect entries into
                // RerankEntriesSet, a HashSet, so nothing depends on the order - and a wide
                // fan-in (Cell.Lift over N cells links all N to one node) makes the repeated
                // RemoveAt(0) that this replaces quadratic in the number of entries.
                List<Entry> entries = this.Node.Entries;
                int last = entries.Count - 1;
                if (index != last)
                {
                    Entry moved = entries[last];
                    entries[index] = moved;
                    moved.nodeEntryIndex = index;
                }

                entries.RemoveAt(last);
            }
        }

        // The general-purpose entry, for the call sites that run once per construction or once
        // per transaction and so have nothing to gain from avoiding the delegate.
        internal sealed class ActionEntry : Entry
        {
            private readonly Action<TransactionInternal> action;

            public ActionEntry(Node node, Action<TransactionInternal> action)
                : base(node) => this.action = action;

            public override void Execute(TransactionInternal trans) => this.action(trans);
        }
    }
}